File: code/300.go
/*
 * @Author: lmio
 * @Date: 2023-04-06 09:35:31
 * @LastEditTime: 2023-08-21 21:42:19
 * @FilePath: /leetcode/code/300.go
 * @Description:300. 最长递增子序列
 */
package code

import "sort"

func LengthOfLIS(nums []int) int {
	stack := []int{nums[0]}
	for _, num := range nums[1:] {
		if num > stack[len(stack)-1] {
			stack = append(stack, num)
		} else {
			idx := sort.SearchInts(stack, num)
			stack[idx] = num
		}
	}
	return len(stack)
}

File: code/300.py
'''
Author: lmio 2091319361@qq.com
Date: 2023-08-21 21:27:16
LastEditors: lmio 2091319361@qq.com
Description: 300. 最长递增子序列
'''

import bisect
from typing import List


class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        stack = [nums[0]]
        for i in range(1, len(nums)):
            num = nums[i]
            if num > stack[-1]:
                stack.append(num)
            else:
                idx = bisect.bisect_left(stack, num)
                stack[idx] = num
        return len(stack)